File: bench/Foo.java
/*
 * Modification History
 *
 * 2001-May-18   Jason Rohrer
 * Created.
 */


/**
 * A simple, naive prime-finding benchmark.
 *
 * @author Jason Rohrer
 */
public class Foo {

		public static void main( String argv[] ) {

			if( argv.length != 1 ) {
				usage( "Foo" );
				}
			
			int limit = Integer.parseInt( argv[0] );

			int numPrimes = 0;
			
			long timeStart = System.currentTimeMillis();

			
			for( int i=5; i<limit; i++ ) {
				if( isPrime(i) ) {
					numPrimes++;
					}
				}
			
			long timeEnd = System.currentTimeMillis();
			long netTime = timeEnd - timeStart;
			System.out.println( "Running time = " + netTime );

			System.out.println( "Number of primes found = " +
								numPrimes );
			}

		

		/**
		 * A naive prime detector.
		 *
		 * @param inNumToCheck the number to check for primality.
		 *
		 * @return ture iff inNumToCheck is prime.
		 */
		public static boolean isPrime( int inNumToCheck ) {
			for( int i=2; i<inNumToCheck; i++ ) {
				if( inNumToCheck % i == 0 ) {
					return false;
					}
				}
			return true;
			}


		
		/**
		 * Prints a usage message and exits.
		 *
		 * @param inAppName the name of the application.
		 */
		public static void usage( String inAppName ) {
	
			System.out.println( "usage:" );
			System.out.println( inAppName + " prime_limit" );
			
			System.out.println( "example:" );
			System.out.println( "\t" + inAppName + " 10000" );
			
			System.exit( 1 );
			}
		}

File: bench/Foo.cpp
/*
 * Modification History
 *
 * 2002-February-13   Jason Rohrer
 * Created.  Copied from C version.
 *
 * 2002-February-14   Jason Rohrer
 * Added a remainder function call to the inner loop.
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <sys/timeb.h>


// prototypes:
void usage( char *inAppName );

// a brute-force, naive prime detector
char isPrime( int n );

// computes the remainder of a division
int remainder( int quotient, int divisor );

// a computationally-intense test application
int main( char inNumArgs, char **inArgs ) {

	int limit;
	struct timeb timeStart;
	struct timeb timeEnd;
	int deltaMS;
	int deltaS;
	
	int i;
	int numPrimes;

	if( inNumArgs != 2 ) {
		usage( inArgs[0] );
		}
	
	if( 1 != sscanf( inArgs[1], "%d", &limit ) ) {
		usage( inArgs[0] );
		}

	numPrimes = 0;

	
	// TIMED portion starts here
	ftime( &timeStart );
	
	for( i=5; i<limit; i++ ) {
		if( isPrime(i) ) {
			numPrimes++;
			}
		}
	
	ftime( &timeEnd );
	// TIMED portion ends here

	
	deltaMS = timeEnd.millitm - timeStart.millitm;
	deltaS = timeEnd.time - timeStart.time;

	// carry, if needed
	if( deltaMS < 0 ) {
		deltaS--;
		deltaMS += 1000;
		}
	
	printf( "Running time = %d\n", 1000 * deltaS + deltaMS );

	printf( "Number of primes found = %d\n", numPrimes );
	}



char isPrime( int inNumToCheck) {
	int i;
	
	for( i=2; i<inNumToCheck; i++ ) {
		if( remainder( inNumToCheck, i ) == 0 ) {
			return 0;
			}
		}
	return 1;
	}



int remainder( int quotient, int divisor ) {
    return quotient % divisor;
    }



void usage( char *inAppName ) {
	
	printf( "usage:\n" );
	printf( "%s prime_limit\n", inAppName );

	printf( "example:\n\t" );
	printf( "%s 10000\n", inAppName );
	
	exit( 1 );
	}
